maint/mocked/based.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2019 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * SPDX-License-Identifier: GPL-2.0-or-later
```

Comment:
My comment about SPDX was meant in general, not just for the one file -- please update all the files to use our standard notice. Again, these are human-readable comments, not intended for machine parsing, and should be readable in natural language.

Reply:
It's not progress, it's obfuscation. Again, this is a comment meant for humans to read. The goal is not "perfect sense" according to some arbitrary formalism, the goal is natural, conversational understanding. "This source code is licensed under ..." is ideally suited to its purpose.

Reply:
It the same song all over, worries about the general human audience stepping
into the highly technical territory, and question whether we shall be guides
for them (see dot files discussion).  Your typical repair shop will not give
you exact details how they fixed your car together with mechanic engineering
intro, this barrier is self-understood and you gotta pay your own effort
to learn such field-specific knowledge.

I claim that it's OK to offer a limited guidance for someone already in
capacity to bootstrap their own exploration from there, but otherwse,
grasping some very common software project structures is up to the
intererested folks to learn on their discretion (guides on C coding, build
systems, CI, etc.).  But it's not our job to do that, it's an expression of
good will at best, and is a way to attract contributions, which also
counts, no doubts, but it needs to be a balanced proposition.
Going into non-sensical depths is really not appealing there.

So, the licensing info for your general human audience is in `COPYING`,
end of story.  The rest is intended for those:

- in posssession of common FOSS projects patterns knowledge, or at least

- in possession of the minimal internet searching capabilities, perhaps
  not even necessarily in English (for better accessibility, I thereby
  propose all the license headers are, applying your argument to show
  how mislead it is, doubled in Czech!)

What I want to show is that your judgement seems at sea to me.

We already use that license annotations; please, let's stop this
bikeshedding discussion and carry on the productive parts instead.
I can pull the respective `COPYING` file addition from other PR
if you want to have this formally backed.

Reply:
The point is not whether the reader is technically proficient or not, it's whether they're *human* or not.

Comments are intended to communicate with humans. Natural language is the technically superior tool for this purpose.

SPDX adds *nothing* to the purpose of the comment, which is to convey to human beings the license that applies to the file. It is an unnecessary level of indirection beyond *just telling them*.

This isn't bike-shedding, it's clearly documented upstream policy.

Reply:
SPDX fulfills exactly the purpose of indicating the file license
unambiguously to _whoever/whatever needs it_, in the least effortful
way possible.

Do you think people will spend their precious time reading such
notices word by word?  I have news for you, nope.  They will just
skim to find the expected keywords.  SPDX alleviates that avoidable
mental burden.  Huge and/or important projects like Linux kernel
and systemd benefit from such simplicity for some time already,
and there was hardly a dispute you are opening here around that.
It's a good idea in general, and that's why several files in the
existing codebase follow that as well already.

> This isn't bike-shedding, it's clearly documented upstream policy.

That's why I am offering to amend `COPYING` respectively, also to
rectify the current de facto usage.  See our "copyright headers"
dispute, your too attached to literal meanings even if they are
not factually aligned.  `Pacemaker Development` likely too, anything
else?

Reply:
I disagree vehemently. SPDX is not least effort; "This file is licensed under" requires only a basic understanding of the English language, whereas an SPDX identifier requires prior knowledge, investigation, or best guess at its full meaning.

Further, this does not even use SPDX correctly. The SPDX website is very clear that the SPDX identifier should be in its own comment, not part of another comment, and, quote, "when a file already contains a standard header or other license notice, the SPDX project recommends that those existing notices should not be removed. The SPDX ID is recommended to be used to supplement, not replace, existing notices in files."

I intend to eventually remove existing references to SPDX, rather than change upstream policy to allow it. If, in the future, SPDX becomes a customary means of providing machine-readable license information in source files, we can re-evaluate adding it; but in no case will it replace the human-readable standard notice.

Reply:
Bad use is fixable, good that you point this out.

But fixing is a completely different thing to reverting to the times of
tarball-only code sharing, those time are long gone, please, let's
live in present.

Reply:
I do not foresee SPDX notices in source files becoming important in either the present or future. SPDX makes sense in places like spec files. While machine-parseable license information within source files might one day become a standard practice, I think the chances of that are very low. If that does happen, we can re-evaluate adding it, but again, it wouldn't replace our natural-language notice.

Reply:
The fact that some people put SPDX in source files now does not make it useful. There isn't enough (or any) benefit to the pacemaker project to clutter source files with it right now.

Reply:
Double checked your claims

> Further, this does not even use SPDX correctly. The SPDX website is
> very clear that the SPDX identifier should be in its own comment,
> not part of another comment

where can I find it?

I see the contrary information, e.g.
https://spdx.org/sites/cpstandard/files/pages/files/using_spdx_license_list_short_identifiers.pdf#page=6

> and, quote, "when a file already contains a standard header or other
> license notice, the SPDX project recommends that those existing notices
> should not be removed. The SPDX ID is recommended to be used to
> supplement, not replace, existing notices in files."

I responded to that already:

> I mean not forcibly, but for new files, it shall be absolutely
> acceptable.

Apparently, for new files, there are no existing notices beforehand.

Moreover, `WITHOUT ANY WARRANTY` tag is again, just another
unnecessary cognitive load[1] -- just as not seeing whether "or later"
applies from the casual first sight (as opposed to SPDX identifiers).


As mentioned, I sadly don't have enough of autonomy to overrule such
bikeshedding decisions even on my own sand, but do note that those
arguments above actually appear rather phony.  I'd greatly appreciate
if you at least kept existing usages as they were and not stir the
git history just for the sake of stirring it.


[1] https://stitcher.io/blog/a-programmers-cognitive-load

> [...] I can safely say that about 80-90% of comments are redundant.
> They should be removed to clear visual overload [...]

Reply:
> Double checked your claims
> 
> > Further, this does not even use SPDX correctly. The SPDX website is
> > very clear that the SPDX identifier should be in its own comment,
> > not part of another comment
> 
> where can I find it?

I looked at the spdx.org site. All of the examples I saw had the ID in a one-line comment of its own, separate from other comments.

> I see the contrary information, e.g.
> https://spdx.org/sites/cpstandard/files/pages/files/using_spdx_license_list_short_identifiers.pdf#page=6

I missed that one.
 
> > and, quote, "when a file already contains a standard header or other
> > license notice, the SPDX project recommends that those existing notices
> > should not be removed. The SPDX ID is recommended to be used to
> > supplement, not replace, existing notices in files."
> 
> I responded to that already:
> 
> > I mean not forcibly, but for new files, it shall be absolutely
> > acceptable.
> 
> Apparently, for new files, there are no existing notices beforehand.

In my mind, I was thinking about existing *policy* rather than just particular existing files. I should have been clearer.
 
> Moreover, `WITHOUT ANY WARRANTY` tag is again, just another
> unnecessary cognitive load[1] -- just as not seeing whether "or later"
> applies from the casual first sight (as opposed to SPDX identifiers).
> 
> As mentioned, I sadly don't have enough of autonomy to overrule such
> bikeshedding decisions even on my own sand, but do note that those

Correct, you cannot unilaterally override upstream policy. Upstream policy is clear and reasonable in this case. If I were submitting a kernel patch, I'd follow kernel style guidelines, and I'd expect to be rejected if I didn't.

Further, slipping in policy changes with unrelated pull requests just bogs down the whole process. If you disagree with particular upstream policy, put in a separate pull request (e.g. against Pacemaker Development), and we can discuss it there, and invite mailing list comment if appropriate.

It really helps to keep PRs focused on the particular issues they are intended to solve and put secondary issues in their own PRs.

> arguments above actually appear rather phony. I'd greatly appreciate
> if you at least kept existing usages as they were and not stir the
> git history just for the sake of stirring it.

My long-term goal is for all source files to meet current style guidelines. I don't want to create a bunch of large changes at one time though for something meta like that.
 
> [1] https://stitcher.io/blog/a-programmers-cognitive-load
> 
> > [...] I can safely say that about 80-90% of comments are redundant.
> > They should be removed to clear visual overload [...]